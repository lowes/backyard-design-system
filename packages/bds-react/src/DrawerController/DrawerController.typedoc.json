{
	"id": 0,
	"name": "@lowes-tech/bds-react/DrawerController",
	"kind": 1,
	"kindString": "Project",
	"flags": {},
	"children": [
		{
			"id": 45,
			"name": "default",
			"kind": 16777216,
			"kindString": "Reference",
			"flags": {},
			"target": 1
		},
		{
			"id": 2,
			"name": "DrawerControllerProps",
			"kind": 256,
			"kindString": "Interface",
			"flags": {},
			"children": [
				{
					"id": 12,
					"name": "OverlayComponent",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Overlay component to display behind modal"
					},
					"type": {
						"type": "reference",
						"typeArguments": [
							{
								"type": "intrinsic",
								"name": "any"
							}
						],
						"name": "ElementType"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.OverlayComponent"
					}
				},
				{
					"id": 3,
					"name": "anchor",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Anchor point for the `Drawer`",
						"text": "Automatically sets `Drawer`'s `orientation` prop if not specified.\n",
						"tags": [
							{
								"tag": "default",
								"text": "'left'\n"
							}
						]
					},
					"type": {
						"type": "union",
						"types": [
							{
								"type": "literal",
								"value": "left"
							},
							{
								"type": "literal",
								"value": "right"
							},
							{
								"type": "literal",
								"value": "top"
							},
							{
								"type": "literal",
								"value": "bottom"
							}
						]
					}
				},
				{
					"id": 8,
					"name": "as",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"type": {
						"type": "reference",
						"typeArguments": [
							{
								"type": "intrinsic",
								"name": "any"
							}
						],
						"name": "ElementType"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.as"
					}
				},
				{
					"id": 11,
					"name": "children",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Children"
					},
					"type": {
						"type": "reference",
						"typeArguments": [
							{
								"type": "intrinsic",
								"name": "any"
							},
							{
								"type": "union",
								"types": [
									{
										"type": "intrinsic",
										"name": "string"
									},
									{
										"type": "reference",
										"typeArguments": [
											{
												"type": "intrinsic",
												"name": "any"
											}
										],
										"name": "JSXElementConstructor"
									}
								]
							}
						],
						"name": "ReactElement"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.children"
					}
				},
				{
					"id": 9,
					"name": "className",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "DOM Class Name"
					},
					"type": {
						"type": "intrinsic",
						"name": "string"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.className"
					}
				},
				{
					"id": 14,
					"name": "container",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Container object or function of modal to display"
					},
					"type": {
						"type": "reference",
						"name": "ReactInstance"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.container"
					}
				},
				{
					"id": 40,
					"name": "context",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Props to provide context for"
					},
					"type": {
						"type": "reference",
						"typeArguments": [
							{
								"type": "intrinsic",
								"name": "string"
							},
							{
								"type": "intrinsic",
								"name": "any"
							}
						],
						"name": "Record"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.context"
					}
				},
				{
					"id": 15,
					"name": "disableAutoFocus",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not browser will auto focus on modal"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.disableAutoFocus"
					}
				},
				{
					"id": 17,
					"name": "disableEnforceFocus",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not modal enforces focus (traps)"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.disableEnforceFocus"
					}
				},
				{
					"id": 18,
					"name": "disableEscapeKeyDown",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not escape key listener is active"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.disableEscapeKeyDown"
					}
				},
				{
					"id": 16,
					"name": "disableOverlayClick",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not overlay click listener is active"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.disableOverlayClick"
					}
				},
				{
					"id": 19,
					"name": "disablePortal",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not portal is used to bring modal to root DOM"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.disablePortal"
					}
				},
				{
					"id": 20,
					"name": "disableRestoreFocus",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not modal restores focus"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.disableRestoreFocus"
					}
				},
				{
					"id": 21,
					"name": "disableScrollLock",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not scroll lock is disabled"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.disableScrollLock"
					}
				},
				{
					"id": 4,
					"name": "drawer",
					"kind": 1024,
					"kindString": "Property",
					"flags": {},
					"comment": {
						"shortText": "Drawer component to control"
					},
					"type": {
						"type": "reference",
						"typeArguments": [
							{
								"type": "intrinsic",
								"name": "any"
							},
							{
								"type": "union",
								"types": [
									{
										"type": "intrinsic",
										"name": "string"
									},
									{
										"type": "reference",
										"typeArguments": [
											{
												"type": "intrinsic",
												"name": "any"
											}
										],
										"name": "JSXElementConstructor"
									}
								]
							}
						],
						"name": "ReactElement"
					}
				},
				{
					"id": 5,
					"name": "drawerStyle",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Drawer style overrides"
					},
					"type": {
						"type": "reference",
						"name": "CSSProperties"
					}
				},
				{
					"id": 22,
					"name": "hideOverlay",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not overlay is hidden"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.hideOverlay"
					}
				},
				{
					"id": 23,
					"name": "keepMounted",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not modal continues to be mounted when closed\nUseful for certain SEO conditions"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.keepMounted"
					}
				},
				{
					"id": 24,
					"name": "manager",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Class that manages modal state and properties"
					},
					"type": {
						"type": "intrinsic",
						"name": "any"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.manager"
					}
				},
				{
					"id": 39,
					"name": "modalStyle",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Style of the modal"
					},
					"type": {
						"type": "reference",
						"name": "CSSProperties"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.modalStyle"
					}
				},
				{
					"id": 7,
					"name": "open",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Whether or not modal is open"
					},
					"type": {
						"type": "intrinsic",
						"name": "boolean"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.open"
					}
				},
				{
					"id": 13,
					"name": "overlayProps",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Props to extend overlay component with\nUseful if the `style` prop needs to be different"
					},
					"type": {
						"type": "reference",
						"name": "OverlayProps"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.overlayProps"
					}
				},
				{
					"id": 41,
					"name": "override",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Props to override children with"
					},
					"type": {
						"type": "reference",
						"typeArguments": [
							{
								"type": "intrinsic",
								"name": "string"
							},
							{
								"type": "intrinsic",
								"name": "any"
							}
						],
						"name": "Record"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.override"
					}
				},
				{
					"id": 6,
					"name": "position",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Position override for `Drawer` placement"
					},
					"type": {
						"type": "reference",
						"name": "CSSProperties"
					},
					"overwrites": {
						"type": "reference",
						"name": "Omit.position"
					}
				},
				{
					"id": 10,
					"name": "style",
					"kind": 1024,
					"kindString": "Property",
					"flags": {
						"isOptional": true
					},
					"comment": {
						"shortText": "Style of modal controller"
					},
					"type": {
						"type": "reference",
						"name": "CSSProperties"
					},
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.style"
					}
				},
				{
					"id": 28,
					"name": "onClose",
					"kind": 2048,
					"kindString": "Method",
					"flags": {
						"isOptional": true
					},
					"signatures": [
						{
							"id": 29,
							"name": "onClose",
							"kind": 4096,
							"kindString": "Call signature",
							"flags": {},
							"comment": {
								"shortText": "`onClose` trigger function, when modal is closed",
								"tags": [
									{
										"tag": "argument",
										"text": "{Event} event - DOM event"
									},
									{
										"tag": "argument",
										"text": "{string} name - Name of event fired that caused modal to close\n"
									}
								]
							},
							"parameters": [
								{
									"id": 30,
									"name": "event",
									"kind": 32768,
									"kindString": "Parameter",
									"flags": {
										"isOptional": true
									},
									"type": {
										"type": "reference",
										"name": "ModalHandlerEvent"
									}
								},
								{
									"id": 31,
									"name": "name",
									"kind": 32768,
									"kindString": "Parameter",
									"flags": {
										"isOptional": true
									},
									"type": {
										"type": "intrinsic",
										"name": "string"
									}
								}
							],
							"type": {
								"type": "intrinsic",
								"name": "void"
							},
							"inheritedFrom": {
								"type": "reference",
								"name": "Omit.onClose"
							}
						}
					],
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.onClose"
					}
				},
				{
					"id": 34,
					"name": "onEscapeKeyDown",
					"kind": 2048,
					"kindString": "Method",
					"flags": {
						"isOptional": true
					},
					"signatures": [
						{
							"id": 35,
							"name": "onEscapeKeyDown",
							"kind": 4096,
							"kindString": "Call signature",
							"flags": {},
							"comment": {
								"shortText": "`onEscapeKeyDown` trigger function, when escape key is pressed while modal is active",
								"tags": [
									{
										"tag": "argument",
										"text": "{Event} event - DOM event\n"
									}
								]
							},
							"parameters": [
								{
									"id": 36,
									"name": "event",
									"kind": 32768,
									"kindString": "Parameter",
									"flags": {},
									"type": {
										"type": "reference",
										"typeArguments": [
											{
												"type": "reference",
												"name": "Element"
											}
										],
										"name": "KeyboardEvent"
									}
								}
							],
							"type": {
								"type": "intrinsic",
								"name": "void"
							},
							"inheritedFrom": {
								"type": "reference",
								"name": "Omit.onEscapeKeyDown"
							}
						}
					],
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.onEscapeKeyDown"
					}
				},
				{
					"id": 32,
					"name": "onOpen",
					"kind": 2048,
					"kindString": "Method",
					"flags": {
						"isOptional": true
					},
					"signatures": [
						{
							"id": 33,
							"name": "onOpen",
							"kind": 4096,
							"kindString": "Call signature",
							"flags": {},
							"comment": {
								"shortText": "`onOpen` trigger function, when modal is closed"
							},
							"type": {
								"type": "intrinsic",
								"name": "void"
							},
							"inheritedFrom": {
								"type": "reference",
								"name": "Omit.onOpen"
							}
						}
					],
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.onOpen"
					}
				},
				{
					"id": 25,
					"name": "onOverlayClick",
					"kind": 2048,
					"kindString": "Method",
					"flags": {
						"isOptional": true
					},
					"signatures": [
						{
							"id": 26,
							"name": "onOverlayClick",
							"kind": 4096,
							"kindString": "Call signature",
							"flags": {},
							"comment": {
								"shortText": "`onOverlayClick` trigger function, when overlay is clicked",
								"tags": [
									{
										"tag": "argument",
										"text": "{Event} event - DOM event\n"
									}
								]
							},
							"parameters": [
								{
									"id": 27,
									"name": "event",
									"kind": 32768,
									"kindString": "Parameter",
									"flags": {},
									"type": {
										"type": "reference",
										"typeArguments": [
											{
												"type": "reference",
												"name": "Element"
											},
											{
												"type": "reference",
												"name": "MouseEvent"
											}
										],
										"name": "MouseEvent"
									}
								}
							],
							"type": {
								"type": "intrinsic",
								"name": "void"
							},
							"inheritedFrom": {
								"type": "reference",
								"name": "Omit.onOverlayClick"
							}
						}
					],
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.onOverlayClick"
					}
				},
				{
					"id": 37,
					"name": "onRendered",
					"kind": 2048,
					"kindString": "Method",
					"flags": {
						"isOptional": true
					},
					"signatures": [
						{
							"id": 38,
							"name": "onRendered",
							"kind": 4096,
							"kindString": "Call signature",
							"flags": {},
							"comment": {
								"shortText": "`onRendered` trigger function, when modal is closed"
							},
							"type": {
								"type": "intrinsic",
								"name": "void"
							},
							"inheritedFrom": {
								"type": "reference",
								"name": "Omit.onRendered"
							}
						}
					],
					"inheritedFrom": {
						"type": "reference",
						"name": "Omit.onRendered"
					}
				}
			],
			"groups": [
				{
					"title": "Properties",
					"kind": 1024,
					"children": [
						12,
						3,
						8,
						11,
						9,
						14,
						40,
						15,
						17,
						18,
						16,
						19,
						20,
						21,
						4,
						5,
						22,
						23,
						24,
						39,
						7,
						13,
						41,
						6,
						10
					]
				},
				{
					"title": "Methods",
					"kind": 2048,
					"children": [
						28,
						34,
						32,
						25,
						37
					]
				}
			],
			"extendedTypes": [
				{
					"type": "reference",
					"typeArguments": [
						{
							"type": "reference",
							"name": "ModalControllerProps"
						},
						{
							"type": "reference",
							"name": "DrawerControllerOverrideProps"
						}
					],
					"name": "Omit"
				}
			]
		},
		{
			"id": 42,
			"name": "DrawerControllerRef",
			"kind": 4194304,
			"kindString": "Type alias",
			"flags": {},
			"type": {
				"type": "reference",
				"name": "ModalControllerRef"
			}
		},
		{
			"id": 1,
			"name": "DrawerController",
			"kind": 32,
			"kindString": "Variable",
			"flags": {},
			"comment": {
				"shortText": "Backyard React Drawer Controller",
				"text": "Extends from `DrawerHandler` to offer and cleaner API to work with\n\nControls drawer defined in `drawer` prop with basic closing/opening handling\n\nFor more API controls integrated with this HOC, see `DrawerHandler`\n\nExample:\n <DrawerController\n     drawer={(\n         <Drawer>\n             <DrawerHeader ... />\n             <DrawerBody ... />\n             <DrawerFooter ... />\n         </Drawer>\n     )}\n />\n"
			},
			"type": {
				"type": "reference",
				"typeArguments": [
					{
						"type": "reference",
						"id": 2,
						"name": "DrawerControllerProps"
					}
				],
				"name": "BDSForwardRef"
			},
			"defaultValue": "..."
		},
		{
			"id": 43,
			"name": "useDrawerController",
			"kind": 64,
			"kindString": "Function",
			"flags": {
				"isConst": true
			},
			"signatures": [
				{
					"id": 44,
					"name": "useDrawerController",
					"kind": 4096,
					"kindString": "Call signature",
					"flags": {},
					"type": {
						"type": "reference",
						"name": "DrawerControllerContextValues"
					}
				}
			]
		}
	],
	"groups": [
		{
			"title": "References",
			"kind": 16777216,
			"children": [
				45
			]
		},
		{
			"title": "Interfaces",
			"kind": 256,
			"children": [
				2
			]
		},
		{
			"title": "Type aliases",
			"kind": 4194304,
			"children": [
				42
			]
		},
		{
			"title": "Variables",
			"kind": 32,
			"children": [
				1
			]
		},
		{
			"title": "Functions",
			"kind": 64,
			"children": [
				43
			]
		}
	]
}