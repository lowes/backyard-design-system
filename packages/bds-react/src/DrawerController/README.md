# @lowes-tech/bds-react/DrawerController

## Interfaces

- [DrawerControllerProps](interfaces/DrawerControllerProps.md)

## References

### default

Renames and re-exports [DrawerController](README.md#drawercontroller)

## Type aliases

### DrawerControllerRef

Ƭ **DrawerControllerRef**: `ModalControllerRef`

## Variables

### DrawerController

• **DrawerController**: `BDSForwardRef`<[`DrawerControllerProps`](interfaces/DrawerControllerProps.md)\>

Backyard React Drawer Controller

Extends from `DrawerHandler` to offer and cleaner API to work with

Controls drawer defined in `drawer` prop with basic closing/opening handling

For more API controls integrated with this HOC, see `DrawerHandler`

Example:
 <DrawerController
     drawer={(
         <Drawer>
             <DrawerHeader ... />
             <DrawerBody ... />
             <DrawerFooter ... />
         </Drawer>
     )}
 />

## Functions

### useDrawerController

▸ `Const` **useDrawerController**(): `DrawerControllerContextValues`

#### Returns

`DrawerControllerContextValues`
