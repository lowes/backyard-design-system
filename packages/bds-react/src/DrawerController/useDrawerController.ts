import * as React from 'react'

import { ModalControllerContext } from '../ModalController/ModalControllerProvider'
import type { ModalControllerContextValues } from '../ModalController/ModalControllerProvider'

const useDrawerController = () => {
    return React.useContext<DrawerControllerContextValues>(ModalControllerContext)
}

interface DrawerControllerContextValues extends ModalControllerContextValues {
    /**
     * Orientation of the `Drawer`
     */
    orientation?: 'vertical' | 'horizontal'
}

export { useDrawerController }

export default useDrawerController
