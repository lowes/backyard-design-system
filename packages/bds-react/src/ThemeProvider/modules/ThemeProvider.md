# Namespace: ThemeProvider

## Variables

### bdsName

â€¢ **bdsName**: `string`
