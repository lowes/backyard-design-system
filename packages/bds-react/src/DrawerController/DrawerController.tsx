import * as React from 'react'
import classNames from 'classnames'
import styled from 'styled-components'

import ModalController from '../ModalController'
import type { ModalControllerProps, ModalControllerRef } from '../ModalController'
import type { BDSForwardRef } from '../utils/typings/BDSComponentProps'
import { useBackyardTheme } from '../ThemeProvider'
import type { DrawerProps } from '../Drawer'

import DrawerControllerBase from './styles/DrawerControllerBase'

// Inherit from `ModalController`
const DrawerControllerWrapper = styled(ModalController)`
    ${DrawerControllerBase}
`

// Maps `DrawerController`'s `anchor` to `Drawer`'s `orientation`
const orientationMap: Record<DrawerControllerProps['anchor'], DrawerProps['orientation']> = {
    top: 'horizontal',
    bottom: 'horizontal',
    left: 'vertical',
    right: 'vertical',
}

// Maps `DraweController`'s `anchor` to CSS Properties for positioning
const positionMap: Record<DrawerControllerProps['anchor'], React.CSSProperties> = {
    top: {
        top: '0',
    },
    bottom: {
        bottom: '0',
    },
    left: {
        left: '0',
    },
    right: {
        right: '0',
    },
}

/**
 * Backyard React Drawer Controller
 *
 * Extends from `DrawerHandler` to offer and cleaner API to work with
 *
 * Controls drawer defined in `drawer` prop with basic closing/opening handling
 *
 * For more API controls integrated with this HOC, see `DrawerHandler`
 *
 * Example:
 *  <DrawerController
 *      drawer={(
 *          <Drawer>
 *              <DrawerHeader ... />
 *              <DrawerBody ... />
 *              <DrawerFooter ... />
 *          </Drawer>
 *      )}
 *  />
 */
const DrawerController: BDSForwardRef<DrawerControllerProps> = React.forwardRef<
    DrawerControllerRef,
    DrawerControllerProps
>(function DrawerController(
    {
        className,
        anchor = 'left',
        drawer: drawerProp,
        style: styleProp,
        drawerStyle: drawerStyleProp,
        position: positionProp,
        ...props
    },
    ref,
) {
    // Get Backyard Theme
    const theme = useBackyardTheme({ validate: true, name: 'DrawerController' })

    // Memoize style prop
    const style = React.useMemo<React.CSSProperties>(
        () => ({
            position: 'fixed',
            top: 0,
            bottom: 0,
            left: 0,
            right: 0,
            zIndex: theme.zIndex.drawer,
            ...styleProp,
        }),
        [theme.name, styleProp],
    )

    // Memoize modal style
    const drawerStyle = React.useMemo<React.CSSProperties>(
        () => ({
            position: 'fixed',
            zIndex: theme.zIndex.drawer + 1,
            transform: 'none',
            ...drawerProp?.props?.style,
            ...drawerStyleProp,
        }),
        [theme.zIndex],
    )

    // Memoize modal render
    const drawer = React.useMemo(
        () =>
            React.cloneElement(drawerProp, {
                style: drawerStyle,
            }),
        [drawerStyle],
    )

    // Memoize position on page
    const position = React.useMemo(
        () => positionProp || positionMap[anchor],
        [positionProp, anchor],
    )

    // Memoize context props
    const context = React.useMemo(
        () => ({
            orientation: orientationMap[anchor],
        }),
        [anchor],
    )

    /**
     * Layout:
     *  <DrawerHandler>
     *      ...
     *  </DrawerHandler>
     */
    return (
        <DrawerControllerWrapper
            className={classNames('drawer-controller', `anchor--${anchor}`, className)}
            modal={drawer}
            modalStyle={drawerStyle}
            position={position}
            context={context}
            style={style}
            {...props}
            ref={ref}
        />
    )
})

type DrawerControllerRef = ModalControllerRef

type DrawerControllerOverrideProps = 'modal'

interface DrawerControllerProps extends Omit<ModalControllerProps, DrawerControllerOverrideProps> {
    /**
     * Anchor point for the `Drawer`
     *
     * Automatically sets `Drawer`'s `orientation` prop if not specified.
     *
     * @default 'left'
     */
    anchor?: 'left' | 'right' | 'top' | 'bottom'
    /**
     * Drawer component to control
     */
    drawer: React.ReactElement
    /**
     * Drawer style overrides
     */
    drawerStyle?: React.CSSProperties
    /**
     * Position override for `Drawer` placement
     */
    position?: React.CSSProperties
}

DrawerController.bdsName = 'DrawerController'

export { DrawerController, DrawerControllerProps, DrawerControllerRef }

export default DrawerController
