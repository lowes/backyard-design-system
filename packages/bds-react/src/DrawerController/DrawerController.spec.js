import * as React from 'react'
import { Settings } from '@lowes-tech/bds-icons'

import { render } from '../../test-utils'
import ThemeProvider from '../ThemeProvider'
import Button from '../Button'
import { Drawer, DrawerBody, DrawerHeader, DrawerFooter } from '../Drawer'

import DrawerController from './DrawerController'

describe('DrawerController Snapshots', () => {
    test('DrawerController default snapshot', () => {
        const { asFragment } = render(
            <DrawerController
                open
                drawer={
                    <Drawer>
                        <DrawerHeader label="Label" icon={<Settings />}>
                            Drawer Heading
                        </DrawerHeader>
                        <DrawerBody>
                            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse
                            cursus fermentum risus, sit amet fringilla nunc pellentesque quis. Duis
                            quis odio ultrices, cursus lacus ac, posuere felis. Donec dignissim
                            libero in augue mattis, a molestie metus vestibulum. Aliquam placerat
                            felis ultrices lorem condimentum, nec ullamcorper felis porttitor.
                        </DrawerBody>
                        <DrawerFooter>
                            <Button variant="secondary">
                                Cancel
                            </Button>
                            <Button>Action</Button>
                        </DrawerFooter>
                    </Drawer>
                }
            />,
        )

        expect(asFragment()).toMatchSnapshot()
    })

    describe('prop: anchor', () => {
        const anchors = ['left', 'right', 'top', 'bottom']

        anchors.forEach((anchor) => {
            test(`DrawerController ${anchor} anchor snapshot`, () => {
                const { asFragment } = render(
                    <DrawerController
                        open
                        anchor={anchor}
                        drawer={
                            <Drawer>
                                <DrawerHeader label="Label" icon={<Settings />}>
                                    Drawer Heading
                                </DrawerHeader>
                                <DrawerBody>
                                    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
                                    Suspendisse cursus fermentum risus, sit amet fringilla nunc
                                    pellentesque quis. Duis quis odio ultrices, cursus lacus ac,
                                    posuere felis. Donec dignissim libero in augue mattis, a
                                    molestie metus vestibulum. Aliquam placerat felis ultrices lorem
                                    condimentum, nec ullamcorper felis porttitor.
                                </DrawerBody>
                                <DrawerFooter>
                                    <Button variant="secondary">
                                        Cancel
                                    </Button>
                                    <Button>Action</Button>
                                </DrawerFooter>
                            </Drawer>
                        }
                    />,
                )

                expect(asFragment()).toMatchSnapshot()
            })
        })
    })

    describe('themes', () => {
        const themes = ['light', 'dark']

        themes.forEach((theme) => {
            test(`DrawerController ${theme} theme snapshot`, () => {
                const { asFragment } = render(
                    <ThemeProvider theme={theme}>
                        <React.Fragment>
                            <DrawerController
                                open
                                drawer={
                                    <Drawer>
                                        <DrawerHeader label="Label" icon={<Settings />}>
                                            Drawer Heading
                                        </DrawerHeader>
                                        <DrawerBody>
                                            Lorem ipsum dolor sit amet, consectetur adipiscing elit.
                                            Suspendisse cursus fermentum risus, sit amet fringilla
                                            nunc pellentesque quis. Duis quis odio ultrices, cursus
                                            lacus ac, posuere felis. Donec dignissim libero in augue
                                            mattis, a molestie metus vestibulum. Aliquam placerat
                                            felis ultrices lorem condimentum, nec ullamcorper felis
                                            porttitor.
                                        </DrawerBody>
                                        <DrawerFooter>
                                            <Button variant="secondary">
                                                Cancel
                                            </Button>
                                            <Button>Action</Button>
                                        </DrawerFooter>
                                    </Drawer>
                                }
                            />
                        </React.Fragment>
                    </ThemeProvider>,
                )

                expect(asFragment()).toMatchSnapshot()
            })
        })
    })

    describe('extra props', () => {
        test('DrawerController DOM props (id, className) snapshot', () => {
            const { asFragment } = render(
                <DrawerController
                    open
                    className="modal-controller-test-class"
                    drawer={
                        <Drawer>
                            <DrawerHeader label="Label" icon={<Settings />}>
                                Drawer Heading
                            </DrawerHeader>
                            <DrawerBody>
                                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse
                                cursus fermentum risus, sit amet fringilla nunc pellentesque quis.
                                Duis quis odio ultrices, cursus lacus ac, posuere felis. Donec
                                dignissim libero in augue mattis, a molestie metus vestibulum.
                                Aliquam placerat felis ultrices lorem condimentum, nec ullamcorper
                                felis porttitor.
                            </DrawerBody>
                            <DrawerFooter>
                                <Button variant="secondary">
                                    Cancel
                                </Button>
                                <Button>Action</Button>
                            </DrawerFooter>
                        </Drawer>
                    }
                />,
            )

            expect(asFragment()).toMatchSnapshot()
        })
    })
})
