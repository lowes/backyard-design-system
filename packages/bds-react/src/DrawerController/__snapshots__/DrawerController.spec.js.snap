// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerController Snapshots DrawerController default snapshot 1`] = `<DocumentFragment />`;

exports[`DrawerController Snapshots extra props DrawerController DOM props (id, className) snapshot 1`] = `<DocumentFragment />`;

exports[`DrawerController Snapshots prop: anchor DrawerController bottom anchor snapshot 1`] = `<DocumentFragment />`;

exports[`DrawerController Snapshots prop: anchor DrawerController left anchor snapshot 1`] = `<DocumentFragment />`;

exports[`DrawerController Snapshots prop: anchor DrawerController right anchor snapshot 1`] = `<DocumentFragment />`;

exports[`DrawerController Snapshots prop: anchor DrawerController top anchor snapshot 1`] = `<DocumentFragment />`;

exports[`DrawerController Snapshots themes DrawerController dark theme snapshot 1`] = `<DocumentFragment />`;

exports[`DrawerController Snapshots themes DrawerController light theme snapshot 1`] = `<DocumentFragment />`;
